dimod/constrained/constrained.py

```diff
@@ -1717,6 +1714,37 @@ def to_file(self, *,
             The ``penalty`` and ``weight`` files, if present, encode the weight
             and penalty type for the constraint.
 
+        Expression Format Specification (Version 2.0):
+
+            This format is inspired by the `NPY format`_
+
+            The first 9 bytes are a magic string: exactly "DIMODEXPR".
+
+            The next 1 byte is an unsigned byte: the major version of the file
+            format.
+
+            The next 1 byte is an unsigned byte: the minor version of the file
+            format.
+
+            The next 4 bytes form a little-endian unsigned int, the length of
+            the header data HEADER_LEN.
+
+            The next HEADER_LEN bytes form the header data. This is a
+            json-serialized dictionary. The dictionary is exactly:
+
+            .. code-block:: python
+
+                data = dict(shape=expr.shape,
+                            dtype=expr.dtype.name,
+                            itype=expr.index_dtype.name,
+                            type=type(expr).__name__,
+                            )
+
+            it is terminated by a newline character and padded with spaces to
+            make the entire length of the entire header divisible by 64.
+
+            The expression data comes after the header.
```

Comment:
This still needs more explanation, but I have simply copy and pasted from the other file to minimize the changes introduced in this PR. Improving docs can come in followup PRs.